src/mlpack/methods/decision_tree/best_binary_numeric_split_impl.hpp

```diff
@@ -185,14 +185,373 @@ double BestBinaryNumericSplit<FitnessFunction>::SplitIfBetter(
+
+  // If we didn't improve, return the original gain exactly as we got it
+  // (without introducing floating point errors).
+  if (!improved)
+    return DBL_MAX;
+
+  if (UseWeights)
+    bestFoundGain /= totalWeight;
+  else
+    bestFoundGain /= data.n_elem;
+
+  return bestFoundGain;
+}
+
+// Optimized version when fitness function is MSEGain.
+template<>
+template<bool UseWeights, typename VecType, typename WeightVecType>
+double BestBinaryNumericSplit<MSEGain>::SplitIfBetter(
+    const double bestGain,
+    const VecType& data,
+    const arma::rowvec& responses,
+    const WeightVecType& weights,
+    const size_t minimumLeafSize,
+    const double minimumGainSplit,
+    double& splitInfo,
+    AuxiliarySplitInfo& /* aux */)
+{
+  // First sanity check: if we don't have enough points, we can't split.
+  if (data.n_elem < (minimumLeafSize * 2))
+    return DBL_MAX;
+  if (bestGain == 0.0)
+    return DBL_MAX; // It can't be outperformed.
+
+  // Next, sort the data.
+  arma::uvec sortedIndices = arma::sort_index(data);
+  arma::rowvec sortedResponses(responses.n_elem);
+  arma::rowvec sortedWeights;
+  for (size_t i = 0; i < sortedResponses.n_elem; ++i)
+    sortedResponses[i] = responses[sortedIndices[i]];
+
+  // Sanity check: if the first element is the same as the last, we can't split
+  // in this dimension.
+  if (data[sortedIndices[0]] == data[sortedIndices[sortedIndices.n_elem - 1]])
+    return DBL_MAX;
+
+  // Only initialize if we are using weights.
+  if (UseWeights)
+  {
+    sortedWeights.set_size(sortedResponses.n_elem);
+    // The weights must keep the same order as the responses.
+    for (size_t i = 0; i < sortedResponses.n_elem; ++i)
+      sortedWeights[i] = weights[sortedIndices[i]];
+  }
+
+  double bestFoundGain = std::min(bestGain + minimumGainSplit, 0.0);
+  bool improved = false;
+  // Force a minimum leaf size of 1 (empty children don't make sense).
+  const size_t minimum = std::max(minimumLeafSize, (size_t) 1);
+
+  double totalWeight = 0.0;
+  double leftChildWeight = 0.0;
+  double rightChildWeight = 0.0;
+  double leftWeightedMean = 0.0;
+  double rightWeightedMean = 0.0;
+  double totalWeightedSumSquares = 0.0;
+  arma::rowvec weightedSumSquares;
+
+  double leftMean = 0.0;
+  double rightMean = 0.0;
+  size_t leftChildSize = 0;
+  size_t rightChildSize = 0;
+  double totalSumSquares = 0.0;
+  arma::rowvec sumSquares;
+
+  // Precomputing prefix sum of squares and prefix weighted sum of squares.
+  // This will be used by MSEGain::Evaluate to efficiently compute gain
+  // values for all possible splits.
+  if (UseWeights)
+  {
+    totalWeight = arma::accu(sortedWeights);
+    bestFoundGain *= totalWeight;
+
+    weightedSumSquares.set_size(data.n_elem);
+    // Stores the weighted sum of squares till the previous index.
+    double prevWeightedSumSquares = 0.0;
+
+    for (size_t i = 0; i < minimum - 1; ++i)
+    {
+      const double w = sortedWeights[i];
+      const double x = sortedResponses[i];
```

Comment:
Yeah. We can do that. Earlier I was under the impression that since this is used internally, it will be sufficient to use arma::rowvec, but I didn't think about the case where responses could be arma::frowvec. I will templateize it 👍